src/compas/files/xml.py

```diff
@@ -100,16 +178,83 @@ def __init__(self, root):
     def from_file(cls, source, tree_parser=None):
         tree_parser = tree_parser or DefaultXMLTreeParser
         tree = ET.parse(source, tree_parser())
+        root = tree.getroot()
+        attach_namespaces(root, source)
         return cls(tree.getroot())
 
     @classmethod
     def from_string(cls, text, tree_parser=None):
         tree_parser = tree_parser or DefaultXMLTreeParser
         root = ET.fromstring(text, tree_parser())
+        source = StringIO(text) if isinstance(text, str) else BytesIO(text)
+        attach_namespaces(root, source)
         return cls(root)
 
 
+class XMLWriter(object):
+    """Writes an XML file from XML object.
+
+    Parameters
+    ----------
+    xml : :class:`compas.files.XML`
+
+    """
+    def __init__(self, xml):
+        self.xml = xml
+
+    def write(self, prettify=False):
+        string = self.to_string(prettify=prettify)
+        with open(self.xml.filepath, 'wb') as f:
+            f.write(string)
+
+    def to_string(self, encoding='utf-8', prettify=False):
+        rough_string = ET.tostring(self.xml.root, encoding=encoding, method='xml')
+        if not prettify:
+            return rough_string
+        return prettify_string(rough_string)
+
+
+class XMLElement(object):
+    def __init__(self, tag, attributes=None, elements=None, text=None):
+        self.tag = tag
+        self.attributes = attributes or {}
+        self.elements = elements or []
+        self.text = text
+
+    def get_root(self):
+        root = ET.Element(self.tag, self.attributes)
+        root.text = self.text
+        return root
+
+    def add_children(self, element):
+        for child in self.elements:
+            subelement = ET.SubElement(element, child.tag, child.attributes)
+            subelement.text = child.text
+            child.add_children(subelement)
+
+
 if compas.is_ironpython():
     from compas.files.xml_cli import CLRXMLTreeParser as DefaultXMLTreeParser
+    from compas.files.xml_cli import prettify_string
+    from compas.files.xml_cli import attach_namespaces
 else:
+    from xml.dom import minidom
+
     DefaultXMLTreeParser = ET.XMLParser
+
+    def prettify_string(rough_string):
+        """Return an XML string with added whitespace for legibility.
+
+        Parameters
+        ----------
+        rough_string : str
+            XML string
+        """
+        reparsed = minidom.parseString(rough_string)
+        return reparsed.toprettyxml(indent="  ", encoding='utf-8')
+
+    def attach_namespaces(root, source):
+        """Parse and find the namespaces declared, and add them to the root's attributes."""
+        namespaces = [node for _, node in ET.iterparse(source, events=['start-ns'])]
+        attrib = {'xmlns:' + ns: uri for ns, uri in namespaces}
+        root.attrib.update(attrib)
```

Comment:
should this not go to the top of the file?

Reply:
Not sure.  @gonzalocasas is there a reason you put it at bottom of the file?

Reply:
No real reason that I remember. Probably leftovers of an early draft.